BUG: Handle overlapping line and bar on the same plot

MartinBraquet opened the pull request:
- [x] closes #61161
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This code
```python
import pandas as pd
import matplotlib.pyplot as plt

index = pd.period_range('2023', periods=3, freq='Y')
df = pd.DataFrame({
    'col1': [10, 20, 30],
    'col2': [40, 25, 10]
}, index=index)

fig, ax = plt.subplots()
df['col1'].plot(kind='bar', ax=ax)
ax2 = ax.twinx()
df['col2'].plot(kind='line', ax=ax2, color = 'r')
plt.show()
```
from the linked issue gives this plot:
![image](https://github.com/user-attachments/assets/9732188c-5478-4e93-ae3c-ae18e77776f3)

And swapping the lines (line before bar)
```python
df['col2'].plot(kind='line', ax=ax2, color = 'r')
df['col1'].plot(kind='bar', ax=ax)
```
gives
![image](https://github.com/user-attachments/assets/35815445-dd3d-4a95-8c27-5a4dd07810be)

I believe the difference of results stems from the fact that the figure is always cropped to match the x-range of the last plot. We can certainly adjust the PR to make the first figure behave like the second one if desired.

Please provide feedback on the overall approach. Once confirmed, I'll expand and cover with tests.

MartinBraquet committed:
Handle line and box on the same plot

MartinBraquet committed:
Handle line and box on the same plot

rhshadrach commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -1868,6 +1870,14 @@ def __init__(
         MPLPlot.__init__(self, data, **kwargs)
 
+        if isinstance(data.index, ABCPeriodIndex):
+            self.ax.freq = data.index.freq
```
Why is the modification of freq necessary? I worry that we don't own `ax`, and that we shouldn't be mutating it. But I'm not too familiar with the plotting code; do we do this anywhere else? I don't see any instances at a glance.

rhshadrach reviewed (changes requested):
Thanks for the PR! Can you add a test in `pandas/tests/plotting/test_series.py`. Make the plots and check that the x-axis is as expected.

MartinBraquet committed:
Refactor PeriodConverter to separate frequency handling.

Introduce `convert_from_freq` method and streamline the conversion process by passing `freq` directly instead of using `axis.freq`. This improves modularity and ensures clearer separation of concerns for frequency handling in Period conversion.

MartinBraquet committed:
Refactor time series plotting logic in Matplotlib backend.

Simplified `tick_pos` calculation by reusing helper methods and added a decorator to register pandas Matplotlib converters in the `_plot` method. This improves clarity and ensures proper integration with the pandas Matplotlib ecosystem.

MartinBraquet committed:
Add test for bar and line plot superposition with same x values

This test ensures that bar and line plots with identical x values are correctly superposed on the same axes. It verifies that the x-tick positions remain consistent across plot types.

MartinBraquet committed:
Merge branch 'main' into plot

MartinBraquet committed:
Fix alignment issue in Series.plot with line and bar.

Resolved a bug that prevented a line and bar from aligning on the same plot in `Series.plot`. This addresses issue #61161 and improves plot consistency when combining these chart types.

MartinBraquet committed:
Refactor time series handling in matplotlib plotting.

Move `x_compat` logic and time series helper methods from `LinePlot` to `MPLPlot` for better reusability and maintainability. This simplifies the `LinePlot` class and centralizes common functionality.

MartinBraquet commented on pandas/plotting/_matplotlib/converter.py:
```diff
@@ -225,16 +225,20 @@ def __call__(self, x, pos: int | None = 0) -> str:
+        return PeriodConverter.convert_from_freq(values, axis.freq)
+
+    @staticmethod
+    def convert_from_freq(values, freq):
```
`convert` only uses the freq attribute of axis, so one should allow the user to pass freq without axis object.

MartinBraquet commented on pandas/plotting/_matplotlib/core.py:
```diff
@@ -288,6 +291,21 @@ def __init__(
 
         self.data = self._ensure_frame(self.data)
 
+        from pandas.plotting import plot_params
+
+        self.x_compat = plot_params["x_compat"]
+        if "x_compat" in self.kwds:
+            self.x_compat = bool(self.kwds.pop("x_compat"))
+
+    @final
+    def _is_ts_plot(self) -> bool:
+        # this is slightly deceptive
+        return not self.x_compat and self.use_index and self._use_dynamic_x()
+
+    @final
+    def _use_dynamic_x(self) -> bool:
+        return use_dynamic_x(self._get_ax(0), self.data)
+
```
Those routines are not exclusive to LinePlot, as they seem related to any type of time series. `_is_ts_plot` is used in this PR to check if a boxplot is a time series.

MartinBraquet commented:
@rhshadrach I covered the change with a test. I believe the PR is now ready for review and then merge.

MartinBraquet committed:
Merge branch 'main' into plot

MartinBraquet committed:
Merge branch 'main' into plot

MartinBraquet committed:
Merge branch 'main' into plot

MartinBraquet commented on doc/source/whatsnew/v3.0.0.rst:
```diff
@@ -762,8 +762,9 @@ Plotting
 - Bug in :meth:`DataFrame.plot.line` raising ``ValueError`` when set both color and a ``dict`` style (:issue:`59461`)
-- Bug in :meth:`DataFrame.plot` that causes a shift to the right when the frequency multiplier is greater than one. (:issue:`57587`)
-- Bug in :meth:`Series.plot` preventing a line and scatter plot from being aligned (:issue:`61005`)
+- Bug in :meth:`DataFrame.plot` that causes a shift to the right when the frequency multiplier is greater than one. (:issue:`57587`
```
```suggestion
- Bug in :meth:`DataFrame.plot` that causes a shift to the right when the frequency multiplier is greater than one. (:issue:`57587`)
```

MartinBraquet committed:
Update doc/source/whatsnew/v3.0.0.rst

MartinBraquet committed:
Apply suggestions from code review

rhshadrach commented on pandas/tests/plotting/test_series.py:
```diff
@@ -971,3 +971,17 @@ def test_secondary_y_subplot_axis_labels(self):
+        s.plot(kind="bar", ax=ax)
+        bar_xticks = ax.get_xticks().tolist()
+        s.plot(kind="line", ax=ax, color="r")
+        line_xticks = ax.get_xticks()[: len(s)].tolist()
```
I'm seeing `ax.get_xticks()` here have the same length as `s`; why have `[: len(s)]` here?

MartinBraquet committed:
Merge branch 'main' into plot

# Conflicts:
#	doc/source/whatsnew/v3.0.0.rst
#	pandas/plotting/_matplotlib/core.py

MartinBraquet committed:
Merge remote-tracking branch 'origin/plot' into plot

MartinBraquet committed:
Fix bar and line plot alignment and x-axis visibility in tests

Ensure bar and line plots share consistent x-axis tick labels and verify that x-axis limits are adjusted to make all plotted elements visible in `test_bar_line_plot` test. These changes improve the robustness of visual alignment and boundary checks.

MartinBraquet committed:
Ensure bar_xticks length matches index in test_series.py

Add an assertion to verify the length of `bar_xticks` aligns with the length of the index. This improves the test's robustness by ensuring the data and ticks remain consistent.

MartinBraquet commented on pandas/tests/plotting/test_series.py:
```diff
@@ -971,3 +971,27 @@ def test_secondary_y_subplot_axis_labels(self):
+        line_xticks = [
+            label for label in ax.get_xticklabels() if label.get_text() in years
+        ]
+        assert len(bar_xticks) == len(index)
+        assert bar_xticks == line_xticks
```
This is checking that 
- the three index years are in the two plots
- the x-tick positions for those years are the same in the two plots

MartinBraquet commented on pandas/tests/plotting/test_series.py:
```diff
@@ -971,3 +971,27 @@ def test_secondary_y_subplot_axis_labels(self):
+        assert len(bar_xticks) == len(index)
+        assert bar_xticks == line_xticks
+        x_limits = ax.get_xlim()
+        assert x_limits[0] <= bar_xticks[0].get_position()[0]
+        assert x_limits[1] >= bar_xticks[-1].get_position()[0]
```
Check that both left-most and right-most data points are shown in the current view

rhshadrach reviewed (approved):
lgtm

MartinBraquet committed:
Merge branch 'main' into plot

MartinBraquet commented:
Only a maintainer is allowed to merge a PR, right?

mroeschke commented:
Thanks @MartinBraquet
